### PR TITLE
Handle underscore variables (_, _foo) in typespecs

### PR DESCRIPTION
**Pull request by slepher**
fix issue
https://github.com/elixir-lang/elixir/issues/8888
and add three test case

### CODE REVIEW

File: lib/elixir/lib/kernel/typespec.ex [hunk 1]
```diff
@@ -962,8 +962,20 @@ defmodule Kernel.Typespec do
 
   defp ensure_no_unused_local_vars!(caller, local_vars) do
     fun = fn
-      {name, :used_once} -> compile_error(caller, "type variable #{name} is unused")
-      _ -> :ok
+      {:_, _used} ->
+        compile_error(caller, "type variable '_' is invalid")
```
Comment (fertapric) [inline lib/elixir/lib/kernel/typespec.ex]: I would move this check to `update_local_vars`

Comment (slepher) [reply lib/elixir/lib/kernel/typespec.ex]: ok, thanks.

File: lib/elixir/test/elixir/typespec_test.exs [hunk 2]
```diff
@@ -128,6 +128,28 @@ defmodule TypespecTest do
       end
     end
 
+    test "type variable unused ignore variable started with _" do
+      test_module do
+        assert @type(foo(_hello) :: integer) == :ok
+      end
+    end
+
+    test "type variable _ should be invalid" do
```
Comment (fertapric) [inline lib/elixir/test/elixir/typespec_test.exs]: Nitpick: I think we can join there two tests

```elixir
    test "type variable _ should be invalid" do
      assert_raise CompileError, ~r"type variable '_' is invalid", fn ->
        test_module do
          @type foo(_) :: integer
        end
      end

      assert_raise CompileError, ~r"type variable '_' is invalid", fn ->
        test_module do
          @type foo(_, _) :: integer
        end
      end
    end
```

File: lib/elixir/test/elixir/typespec_test.exs [hunk 3]
```diff
@@ -128,6 +128,28 @@ defmodule TypespecTest do
       end
     end
 
+    test "type variable unused ignore variable started with _" do
```
Comment (fertapric) [inline lib/elixir/test/elixir/typespec_test.exs]:
```suggestion
    test "@type with a variable starting with underscore" do
```

**Comment by fertapric** [comment]
Thanks @slepher! Overall looks good 👌

File: lib/elixir/lib/kernel/typespec.ex (hunk 1 above)
Comment (slepher) [reply lib/elixir/lib/kernel/typespec.ex]: I noticed ensure_no_unused_local_vars! is also used in translate_spec
it's not works as expected.
I will move this check to new function ensure_no_underscore_local_vars! and use it only in translate_type.

File: lib/elixir/lib/kernel/typespec.ex [hunk 4]
```diff
@@ -960,10 +961,29 @@ defmodule Kernel.Typespec do
     end
   end
 
+  defp ensure_no_underscore_local_vars!(caller, var_names) do
+    case :lists.any(&(&1 == :_), var_names) do
```
Comment (josevalim) [inline lib/elixir/lib/kernel/typespec.ex]:
```suggestion
    case :lists.member(:_, var_names) do
```

**Comment by fertapric** [comment]
Thanks @slepher!

**Pull request merged**